src/gateway/scheduling.py

```diff
@@ -140,57 +154,74 @@ def _process(self):
+        if self._event.isSet():  # If a new schedule is saved, stop hanging and refresh the schedules
+            self._event.clear()
+            return
+        thread = BaseThread(name='schedulingexc',
+                            target=self._execute_schedule, args=(schedules_to_execute, schedule))
+        thread.daemon = True
+        thread.start()
 
     @staticmethod
     def _get_next_execution(schedule_dto):
         # type: (ScheduleDTO) -> Optional[float]
         if schedule_dto.repeat is None:
-            return None
+            # Check if start has passed
+            return schedule_dto.start
         base_time = max(SchedulingController.NO_NTP_LOWER_LIMIT, schedule_dto.start, time.time())
         cron = croniter(schedule_dto.repeat,
                         datetime.fromtimestamp(base_time,
                                                pytz.timezone(SchedulingController.TIMEZONE)))
         return cron.get_next(ret_type=float)
 
-    def _execute_schedule(self, schedule_dto, schedule):
-        # type: (ScheduleDTO, Schedule) -> None
-        if schedule_dto.running:
-            return
-        try:
-            schedule_dto.running = True
-            logger.debug('Executing schedule {0} ({1})'.format(schedule_dto.name, schedule_dto.action))
-            if schedule_dto.arguments is None:
-                raise ValueError('Invalid schedule arguments')
-            # Execute
-            if schedule_dto.action == 'GROUP_ACTION':
-                self._group_action_controller.do_group_action(schedule_dto.arguments)
-            elif schedule_dto.action == 'BASIC_ACTION':
-                self._gateway_api.do_basic_action(**schedule_dto.arguments)
-            elif schedule_dto.action == 'LOCAL_API':
-                func = getattr(self._web_interface, schedule_dto.arguments['name'])
-                func(**schedule_dto.arguments['parameters'])
-            else:
-                logger.warning('Did not process schedule_dto {0}'.format(schedule_dto.name))
-
-            # Cleanup or prepare for next run
-            schedule_dto.last_executed = time.time()
-            if schedule_dto.has_ended:
-                schedule_dto.status = 'COMPLETED'
-                schedule.status = 'COMPLETED'
-                schedule.save()
-        except CommunicationTimedOutException as ex:
-            logger.error('Got error while executing schedule: {0}'.format(ex))
-        except Exception as ex:
-            logger.error('Got error while executing schedule: {0}'.format(ex))
-            schedule_dto.last_executed = time.time()
-        finally:
-            schedule_dto.running = False
-            schedule_dto.next_execution = SchedulingController._get_next_execution(schedule_dto)
+    def _execute_schedule(self, schedules_to_execute, schedule):
+        # type: (List[ScheduleDTO], Schedule) -> None
+        for schedule_dto in schedules_to_execute:
+            if schedule_dto.running:
+                return
```

Comment:
Since this now handles multiple schedules this should only break out of the current loop.

```suggestion
                continue
```

Reply:
nice catch :)